shuup/admin/modules/categories/forms.py

```diff
@@ -0,0 +1,104 @@
+# -*- coding: utf-8 -*-
+# This file is part of Shuup.
+#
+# Copyright (c) 2012-2016, Shoop Ltd. All rights reserved.
+#
+# This source code is licensed under the AGPLv3 license found in the
+# LICENSE file in the root directory of this source tree.
+from django import forms
+from django.utils.translation import ugettext_lazy as _
+
+from shuup.admin.forms.fields import Select2MultipleField
+from shuup.admin.forms.widgets import MediaChoiceWidget
+from shuup.admin.utils.forms import filter_form_field_choices
+from shuup.core.models import Category, CategoryStatus, Product, ShopProduct
+from shuup.utils.multilanguage_model_form import MultiLanguageModelForm
+
+
+class CategoryBaseForm(MultiLanguageModelForm):
+    class Meta:
+        model = Category
+        fields = (
+            "parent",
+            "shops",
+            "status",
+            "ordering",
+            "visibility",
+            "visibility_groups",
+            "name",
+            "image",
+            "description",
+            "slug",
+        )
+
+        widgets = {
+            "status": forms.RadioSelect,
+            "visibility": forms.RadioSelect,
+        }
+
+    def __init__(self, **kwargs):
+        super(CategoryBaseForm, self).__init__(**kwargs)
+        # Exclude `DELETED`. We don't want people to use that field to set a category as deleted.
+        filter_form_field_choices(self.fields["status"], (CategoryStatus.DELETED.value,), invert=True)
+
+        # Exclude current category from parents, because it cannot be its own child anyways
+        filter_form_field_choices(self.fields["parent"], (kwargs["instance"].pk,), invert=True)
+        self.fields["image"].widget = MediaChoiceWidget(clearable=True)
+
+
+class CategoryProductForm(forms.Form):
+    primary_products = Select2MultipleField(
+        label=_("Primary Category"),
+        help_text=_("Set this category as a primary for selected products."),
+        model=Product,
+        required=False)
+    update_product_category = forms.BooleanField(
+        label=_("Default Category"),
+        help_text=_("Update product default category while setting primary category."),
+        required=False)
+    additional_products = Select2MultipleField(
+        label=_("Additional Category"),
+        help_text=_("Add selected products to this category"),
+        model=Product,
+        required=False)
+    remove_products = forms.MultipleChoiceField(
+        label=_("Remove Products"),
+        help_text=_("Remove selected products from this category"),
+        required=False)
+
+    def __init__(self, shop, category, **kwargs):
+        self.shop = shop
+        self.category = category
+        super(CategoryProductForm, self).__init__(**kwargs)
+        self.fields["remove_products"].choices = [(None, "-----")] + [
+            (obj.pk, obj.product.name) for obj in category.shop_products.filter(shop=shop)
+        ]
+
+    def save(self):
```

Comment:
if you make this atomic and do your shop product queries using product_id__in=some_list, this operation will be much faster since atomic batches saves. Might be unnecessary optimization at this point though...